---
title: "4_numbat"
author: "zm"
date: "2024-08-29"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
set.seed(42)
```

```{r here}
library(here)
```

```{r libraries}
library(BiocParallel)
library(SingleCellExperiment)
library(tidyverse)

conflicted::conflict_prefer_all(winner = "dplyr", quiet = TRUE)
```

```{r params}
source(file = here("eval", "scripts", "r", "params.r"))
plots <- list()
```

```{r import}
sce_list <- read_rds(
  file = here("data/intermed/snrnaseq/2_sce_line_list.rds")
)
sce_all <- map(
  .x = sce_list,
  .f = function(sce) {
    # no bueno
    reducedDim(sce, "PCA") <- NULL
    reducedDim(sce, "UMAP") <- NULL

    sce
  }
) |>
  purrr::reduce(.f = cbind)

sce_scvi <- read_rds(
  file =  here("data/intermed/snrnaseq/7_sce_scvi.rds")
)
```

```{r fix_bam, eval=FALSE, include=FALSE}
# the bam doesn't have the __
# EnvironmentModules::module_load("openssl/3.2.1")
# EnvironmentModules::module_load("miniconda3")
# 
# library(reticulate)
# reticulate::use_condaenv(condaenv = "scvi-env")
# 
# map2(
#   .x = c("SCRNA_240430", "SCRNA_240621"),
#   .y = c("__s1", "__s2"),
#   .f = function(x, y) {
#     phyloRNA::bamtagregex(
#       input = here("data/processed", x, "process/barcode_headAligned_anno.bam"),
#       output = here("data/processed", x, "append.bam"),
#       tag = "CB",
#       pattern = "$",
#       replace = y,
#       remake = TRUE
#     )
#   }
# )
```
  
```{r eval=FALSE, include=FALSE}
# samtools view -H data/processed/SCRNA_240430_240621/merged_sorted.bam |\
#    sed -e 's/SN:hg38_1/SN:1/' | sed -e 's/SN:hg38_2/SN:2/' | \
#    sed -e 's/SN:hg38_3/SN:3/' | sed -e 's/SN:hg38_4/SN:4/' | \
#    sed -e 's/SN:hg38_5/SN:5/' | sed -e 's/SN:hg38_6/SN:6/' | \
#    sed -e 's/SN:hg38_7/SN:7/' | sed -e 's/SN:hg38_8/SN:8/' | \
#    sed -e 's/SN:hg38_9/SN:9/' | sed -e 's/SN:hg38_10/SN:10/' | \
#    sed -e 's/SN:hg38_11/SN:11/' | sed -e 's/SN:hg38_12/SN:12/' | \
#    sed -e 's/SN:hg38_13/SN:13/' | sed -e 's/SN:hg38_14/SN:14/' | \
#    sed -e 's/SN:hg38_15/SN:15/' | sed -e 's/SN:hg38_16/SN:16/' | \
#    sed -e 's/SN:hg38_17/SN:17/' | sed -e 's/SN:hg38_18/SN:18/' | \
#    sed -e 's/SN:hg38_19/SN:19/' | sed -e 's/SN:hg38_20/SN:20/' | \
#    sed -e 's/SN:hg38_21/SN:21/' | sed -e 's/SN:hg38_22/SN:22/' | \
#    sed -e 's/SN:hg38_X/SN:X/' | sed -e 's/SN:hg38_Y/SN:Y/' | \
#    sed -e 's/SN:hg38_MT/SN:M/' | samtools reheader - data/processed/SCRNA_240430_240621/merged_sorted.bam > data/processed/SCRNA_240430_240621/merged_sorted_chrfix.bam
#
# phyloRNA::bamtagregex(
#   input = here("data/processed/SCRNA_240430_240621/merged_sorted.bam"),
#   output = here("data/processed/SCRNA_240430_240621/merged_sorted_chromfix.bam"),
#   tag = "CB",
#   pattern = "$",
#   replace = y,
#   remake = TRUE
# )
```

```{r write_barcodes}
iwalk(
  .x = sce_list,
  .f = function(sce, i) {
    colData(sce) |>
      as_tibble() |>
      pull(bc_wells) |>
      as_tibble_col() |>
      write_csv(
        col_names = FALSE,
        file = here("data/processed/SCRNA_240430_240621/numbat/barcodes", paste0(i, "_barcodes.csv"))
      )
  }
)
```

```{r eval=FALSE, include=FALSE}
# bash submit.slurm
```

```{r numbat_res}
# leiden 23 == immune
ref_numbat <- counts(sce_all)

rownames(ref_numbat) <- rowData(sce_all)$gene_name
ref_numbat <- numbat::aggregate_counts(
  count_mat = ref_numbat,
  annot = colData(sce_scvi) |>
    as_tibble() |>
    select(bc_wells, leiden) |>
    filter(leiden == 23) |>
    mutate(leiden = "immune") |>
    rename(cell = bc_wells, group = leiden)
)
ref_numbat <- ref_numbat[!duplicated(rownames(ref_numbat)), ]

imap(
  .progress = TRUE,
  .x = sce_list,
  .f = function(sce, i) {
    library(numbat)

    # prep matrix
    sce$bc_wells_corr <- colData(sce) |>
      as_tibble() |>
      pull(bc_wells)
    count_mat <- counts(sce)
    colnames(count_mat) <- sce$bc_wells_corr
    rownames(count_mat) <- rowData(sce)$gene_name
    count_mat <- count_mat[!duplicated(rownames(count_mat)), ]

    # prep ref snp
    snp_consenseus_path <-
      list(
        GL0028 = "/stornext/Bioinf/data/lab_brain_cancer/data_processed/SNP/plots/GL0028/207686170123_R01C02_segments.tsv",
        GL0038 = "/stornext/Bioinf/data/lab_brain_cancer/data_processed/SNP/plots/GL0038/207686170123_R02C01_segments.tsv",
        GL0095 = "/stornext/Bioinf/data/lab_brain_cancer/data_processed/SNP/plots/GL0095/208115890028_R02C01_segments.tsv",
        GL0128 = "/stornext/Bioinf/data/lab_brain_cancer/data_processed/SNP/plots/GL0128/208579500154_R03C01_segments.tsv"
      )
    snp_consenseus <- read_csv(
      file = snp_consenseus_path[[i]],
      show_col_types = FALSE
    )
    snp_consenseus <- snp_consenseus |>
      mutate(
        cnv_state = case_when(
          nMajor == 1 & nMinor == 1 ~ "neu",
          nMajor == 2 & nMinor == 0 ~ "loh",
          nMajor == 1 & nMinor == 0 | nMajor == 0 & nMinor == 1 ~ "del",
          nMajor + nMinor == 0 ~ "bdel",
          nMajor >= 1 & nMinor == 1 | nMajor == 1 & nMinor >= 1 | nMajor > 3 & nMinor == 0 ~ "amp",
          # redundant?
          nMajor > nMinor & nMinor > 0 | nMinor > nMajor & nMajor > 0 ~ "amp",
          nMajor == nMinor & nMajor >= 2 ~ "bamp"
        )
      ) |>
      rename(
        CHROM = chr,
        seg_start = startpos,
        seg_end = endpos
      ) |>
      group_by(CHROM) |>
      mutate(
        seg = paste(as.character(CHROM), letters[seq_along(CHROM)], sep = "")
      ) |>
      select(
        CHROM, seg, seg_start, seg_end, cnv_state
      )

    gc()

    numbat::run_numbat(
      count_mat = count_mat |> as.matrix(),
      lambdas_ref = ref_numbat,
      df_allele = read_tsv(
        here("data/processed/SCRNA_240430_240621/numbat/out", i, paste0(i, "_allele_counts.tsv.gz")),
        show_col_types = FALSE
      ),
      genome = "hg38",
      out_dir = here("data/processed/SCRNA_240430_240621/numbat/out_final", i),
      # t = 1e-5,
      ncores = 50,
      # max_entropy = 10,
      segs_consensus_fix = snp_consenseus,
      plot = TRUE
    )
  }
)

numbat_res <- map(
  .x = names(sce_list),
  .f = function(x) {
    library(numbat)
    nb <- Numbat$new(out_dir = here("data/processed/SCRNA_240430_240621/numbat/out_final", x))

    nb$clone_post |>
      as_tibble() |>
      rename(bc_wells_corr = cell, numbat_res = compartment_opt) |>
      select(bc_wells_corr, numbat_res, p_cnv) |>
      mutate(line = x) |>
      relocate(line)
  }
) |>
  bind_rows()

write_rds(
  x = numbat_res,
  file = here("data/processed/SCRNA_240430_240621/numbat/final_numbat_res.rds")
)
```

```{r}
numbat_res <- read_rds(
  file = here("data/processed/SCRNA_240430_240621/numbat/final_numbat_res.rds")
)
sce_scvi$numbat_res <- colData(sce_scvi) |>
  as_tibble() |>
  left_join(numbat_res |> rename(bc_wells = bc_wells_corr)) |>
  pull(numbat_res)
sce_scvi$p_cnv <- colData(sce_scvi) |>
  as_tibble() |>
  left_join(numbat_res |> rename(bc_wells = bc_wells_corr)) |>
  pull(p_cnv)
sce_scvi$line <- colData(sce_scvi) |>
  as_tibble() |>
  left_join(numbat_res |> rename(bc_wells = bc_wells_corr)) |>
  pull(line)

scater::plotReducedDim(
  object = sce_scvi[, !sce_scvi$line == "GL0128"],
  dimred = "X_scVI_MDE",
  colour_by = I(sce_scvi[, !sce_scvi$line == "GL0128"]$p_cnv < 0.5),
  # colour_by = I(sce_scvi[, !sce_scvi$line == "GL0128"]$individual_anno)
)
```

```{r filt}
sce_scvi$normal_res <-
  sce_scvi$line == "GL0128" | sce_scvi$numbat_res == "normal" | sce_scvi$leiden %in% c(22, 23, 24)

sce_scvi_filt <- sce_scvi[, !sce_scvi$normal_res]

scater::plotReducedDim(
  object = sce_scvi_filt,
  dimred = "X_scVI_MDE",
  colour_by = "leiden",
  text_by = "leiden"
)
scater::plotReducedDim(
  object = sce_scvi_filt,
  dimred = "X_scVI_MDE",
  colour_by = "individual_anno"
)
```

```{r save}
# reducedDim(sce_scvi_filt, "X_scVI_UMAP") <- scater::calculateUMAP(
#   x = sce_scvi_filt,
#   dimred = "X_scVI"
# )

# this is an iterative process of going back and forth
# between here and jupyter
# scvi clusters still have outliers
bc_well_outliers <- reducedDim(sce_scvi_filt, "X_scVI_MDE") |>
  as_tibble(rownames = "bc_well")
  # filter(V1 > 2) |>
  # pull(bc_well)

sce_scvi_filt <- sce_scvi_filt[, !sce_scvi_filt$bc_wells %in% bc_well_outliers]

colData(sce_all) <- colData(sce_all) |>
  as_tibble() |>
  left_join(
    y = colData(sce_scvi_filt) |>
      as_tibble() |>
      select(bc_wells, leiden)
  ) |>
  DataFrame()
sce_all$leiden_original <- sce_all$leiden

BiocManager::install("zellkonverter", force = TRUE, update = FALSE)
zellkonverter::writeH5AD(
  sce = sce_all[, sce_all$bc_wells %in% sce_scvi_filt$bc_wells],
  file = here("data/intermed/snrnaseq/8_sce_scvi_tumour.h5ad"),
  verbose = TRUE
)
```
