---
title: "4_numbat"
author: "zm"
date: "2024-08-29"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
set.seed(42)
```

```{r here}
library(here)
```

```{r libraries}
library(BiocParallel)
library(SingleCellExperiment)
library(tidyverse)

conflicted::conflict_prefer_all(winner = "dplyr", quiet = TRUE)
```

```{r params}
source(file = here("eval", "scripts", "r", "params.r"))
plots <- list()
```

```{r import}
sce_list <- read_rds(
  file = here("data/intermed/snrnaseq/2_sce_line_list.rds")
)
sce_all <- map(
  .x = sce_list,
  .f = function(sce) {
    # no bueno
    reducedDim(sce, "PCA") <- NULL
    reducedDim(sce, "UMAP") <- NULL

    sce
  }
) |>
  purrr::reduce(.f = cbind)

sce_scvi <- read_rds(
  file =  here("data/intermed/snrnaseq/7_sce_scvi.rds")
)
```

```{r fix_bam, eval=FALSE, include=FALSE}
# the bam doesn't have the __
# EnvironmentModules::module_load("openssl/3.2.1")
# EnvironmentModules::module_load("miniconda3")
# 
# library(reticulate)
# reticulate::use_condaenv(condaenv = "scvi-env")
# 
# map2(
#   .x = c("SCRNA_240430", "SCRNA_240621"),
#   .y = c("__s1", "__s2"),
#   .f = function(x, y) {
#     phyloRNA::bamtagregex(
#       input = here("data/processed", x, "process/barcode_headAligned_anno.bam"),
#       output = here("data/processed", x, "append.bam"),
#       tag = "CB",
#       pattern = "$",
#       replace = y,
#       remake = TRUE
#     )
#   }
# )
```
  
```{r eval=FALSE, include=FALSE}
# samtools view -H data/processed/SCRNA_240430_240621/merged_sorted.bam |\
#    sed -e 's/SN:hg38_1/SN:1/' | sed -e 's/SN:hg38_2/SN:2/' | \
#    sed -e 's/SN:hg38_3/SN:3/' | sed -e 's/SN:hg38_4/SN:4/' | \
#    sed -e 's/SN:hg38_5/SN:5/' | sed -e 's/SN:hg38_6/SN:6/' | \
#    sed -e 's/SN:hg38_7/SN:7/' | sed -e 's/SN:hg38_8/SN:8/' | \
#    sed -e 's/SN:hg38_9/SN:9/' | sed -e 's/SN:hg38_10/SN:10/' | \
#    sed -e 's/SN:hg38_11/SN:11/' | sed -e 's/SN:hg38_12/SN:12/' | \
#    sed -e 's/SN:hg38_13/SN:13/' | sed -e 's/SN:hg38_14/SN:14/' | \
#    sed -e 's/SN:hg38_15/SN:15/' | sed -e 's/SN:hg38_16/SN:16/' | \
#    sed -e 's/SN:hg38_17/SN:17/' | sed -e 's/SN:hg38_18/SN:18/' | \
#    sed -e 's/SN:hg38_19/SN:19/' | sed -e 's/SN:hg38_20/SN:20/' | \
#    sed -e 's/SN:hg38_21/SN:21/' | sed -e 's/SN:hg38_22/SN:22/' | \
#    sed -e 's/SN:hg38_X/SN:X/' | sed -e 's/SN:hg38_Y/SN:Y/' | \
#    sed -e 's/SN:hg38_MT/SN:M/' | samtools reheader - data/processed/SCRNA_240430_240621/merged_sorted.bam > data/processed/SCRNA_240430_240621/merged_sorted_chrfix.bam
#
# phyloRNA::bamtagregex(
#   input = here("data/processed/SCRNA_240430_240621/merged_sorted.bam"),
#   output = here("data/processed/SCRNA_240430_240621/merged_sorted_chromfix.bam"),
#   tag = "CB",
#   pattern = "$",
#   replace = y,
#   remake = TRUE
# )
```

```{r write_barcodes}
iwalk(
  .x = sce_list,
  .f = function(sce, i) {
    colData(sce) |>
      as_tibble() |>
      pull(bc_wells) |>
      as_tibble_col() |>
      write_csv(
        col_names = FALSE,
        file = here("data/processed/SCRNA_240430_240621/numbat/barcodes", paste0(i, "_barcodes.csv"))
      )
  }
)
```

```{r eval=FALSE, include=FALSE}
# bash submit.slurm
```

```{r numbat_res}
# leiden 23 == immune
ref_numbat <- counts(sce_all)

rownames(ref_numbat) <- rowData(sce_all)$gene_name
ref_numbat <- numbat::aggregate_counts(
  count_mat = ref_numbat,
  annot = colData(sce_scvi) |>
    as_tibble() |>
    select(bc_wells, leiden) |>
    filter(leiden == 23) |>
    mutate(leiden = "immune") |>
    rename(cell = bc_wells, group = leiden)
)
ref_numbat <- ref_numbat[!duplicated(rownames(ref_numbat)), ]

imap(
  .progress = TRUE,
  .x = sce_list,
  .f = function(sce, i) {
    library(numbat)

    # prep matrix
    sce$bc_wells_corr <- colData(sce) |>
      as_tibble() |>
      pull(bc_wells)
    count_mat <- counts(sce)
    colnames(count_mat) <- sce$bc_wells_corr
    rownames(count_mat) <- rowData(sce)$gene_name
    count_mat <- count_mat[!duplicated(rownames(count_mat)), ]

    # prep ref snp
    snp_consenseus_path <-
      list(
        GL0028 = "/stornext/Bioinf/data/lab_brain_cancer/data_processed/SNP/plots/GL0028/207686170123_R01C02_segments.tsv",
        GL0038 = "/stornext/Bioinf/data/lab_brain_cancer/data_processed/SNP/plots/GL0038/207686170123_R02C01_segments.tsv",
        GL0095 = "/stornext/Bioinf/data/lab_brain_cancer/data_processed/SNP/plots/GL0095/208115890028_R02C01_segments.tsv",
        GL0128 = "/stornext/Bioinf/data/lab_brain_cancer/data_processed/SNP/plots/GL0128/208579500154_R03C01_segments.tsv"
      )
    snp_consenseus <- read_csv(
      file = snp_consenseus_path[[i]],
      show_col_types = FALSE
    )
    snp_consenseus <- snp_consenseus |>
      mutate(
        cnv_state = case_when(
          nMajor == 1 & nMinor == 1 ~ "neu",
          nMajor == 2 & nMinor == 0 ~ "loh",
          nMajor == 1 & nMinor == 0 | nMajor == 0 & nMinor == 1 ~ "del",
          nMajor + nMinor == 0 ~ "bdel",
          nMajor >= 1 & nMinor == 1 | nMajor == 1 & nMinor >= 1 | nMajor > 3 & nMinor == 0 ~ "amp",
          # redundant?
          nMajor > nMinor & nMinor > 0 | nMinor > nMajor & nMajor > 0 ~ "amp",
          nMajor == nMinor & nMajor >= 2 ~ "bamp"
        )
      ) |>
      rename(
        CHROM = chr,
        seg_start = startpos,
        seg_end = endpos
      ) |>
      group_by(CHROM) |>
      mutate(
        seg = paste(as.character(CHROM), letters[seq_along(CHROM)], sep = "")
      ) |>
      select(
        CHROM, seg, seg_start, seg_end, cnv_state
      )

    gc()

    numbat::run_numbat(
      count_mat = count_mat |> as.matrix(),
      lambdas_ref = ref_numbat,
      df_allele = read_tsv(
        here("data/processed/SCRNA_240430_240621/numbat/out", i, paste0(i, "_allele_counts.tsv.gz")),
        show_col_types = FALSE
      ),
      genome = "hg38",
      out_dir = here("data/processed/SCRNA_240430_240621/numbat/out_final", i),
      # t = 1e-5,
      ncores = 50,
      # max_entropy = 10,
      segs_consensus_fix = snp_consenseus,
      plot = TRUE
    )
  }
)

numbat_res <- map(
  .x = names(sce_list),
  .f = function(x) {
    library(numbat)
    nb <- Numbat$new(out_dir = here("data/processed/SCRNA_240430_240621/numbat/out_final", x))

    nb$clone_post |>
      as_tibble() |>
      rename(bc_wells_corr = cell, numbat_res = compartment_opt) |>
      select(bc_wells_corr, numbat_res, p_cnv) |>
      mutate(line = x) |>
      relocate(line)
  }
) |>
  bind_rows()

write_rds(
  x = numbat_res,
  file = here("data/processed/SCRNA_240430_240621/numbat/final_numbat_res.rds")
)
```

```{r}
numbat_res <- read_rds(
  file = here("data/processed/SCRNA_240430_240621/numbat/final_numbat_res.rds")
)
sce_scvi$numbat_res <- colData(sce_scvi) |>
  as_tibble() |>
  left_join(numbat_res |> rename(bc_wells = bc_wells_corr)) |>
  pull(numbat_res)
sce_scvi$p_cnv <- colData(sce_scvi) |>
  as_tibble() |>
  left_join(numbat_res |> rename(bc_wells = bc_wells_corr)) |>
  pull(p_cnv)
sce_scvi$line <- colData(sce_scvi) |>
  as_tibble() |>
  left_join(numbat_res |> rename(bc_wells = bc_wells_corr)) |>
  pull(line)

scater::plotReducedDim(
  object = sce_scvi[, !sce_scvi$line == "GL0128"],
  dimred = "X_scVI_MDE",
  colour_by = I(sce_scvi[, !sce_scvi$line == "GL0128"]$p_cnv < 0.5),
  # colour_by = I(sce_scvi[, !sce_scvi$line == "GL0128"]$individual_anno)
)
```

```{r filt}
sce_scvi$normal_res <-
  sce_scvi$line == "GL0128" | sce_scvi$numbat_res == "normal" | sce_scvi$leiden %in% c(22, 23, 24)

sce_scvi_filt <- sce_scvi[, !sce_scvi$normal_res]

scater::plotReducedDim(
  object = sce_scvi_filt,
  dimred = "X_scVI_MDE",
  colour_by = "leiden",
  text_by = "leiden"
)
scater::plotReducedDim(
  object = sce_scvi_filt,
  dimred = "X_scVI_MDE",
  colour_by = "individual_anno"
)
```

```{r save}
# reducedDim(sce_scvi_filt, "X_scVI_UMAP") <- scater::calculateUMAP(
#   x = sce_scvi_filt,
#   dimred = "X_scVI"
# )

# this is an iterative process of going back and forth
# between here and jupyter
# scvi clusters still have outliers
bc_well_outliers <- reducedDim(sce_scvi_filt, "X_scVI_MDE") |>
  as_tibble(rownames = "bc_well")
  # filter(V1 > 2) |>
  # pull(bc_well)

sce_scvi_filt <- sce_scvi_filt[, !sce_scvi_filt$bc_wells %in% bc_well_outliers]

colData(sce_all) <- colData(sce_all) |>
  as_tibble() |>
  left_join(
    y = colData(sce_scvi_filt) |>
      as_tibble() |>
      select(bc_wells, leiden)
  ) |>
  DataFrame()
sce_all$leiden_original <- sce_all$leiden

BiocManager::install("zellkonverter", force = TRUE, update = FALSE)
zellkonverter::writeH5AD(
  sce = sce_all[, sce_all$bc_wells %in% sce_scvi_filt$bc_wells],
  file = here("data/intermed/snrnaseq/8_sce_scvi_tumour.h5ad"),
  verbose = TRUE
)
```





<!-- ```{r} -->
<!-- EnvironmentModules::module_load("texlive") -->

<!-- sce_scvi_tumour <- read_rds("/vast/scratch/users/moore.z/pdo_compar/data/intermed/snrnaseq/9_tumour_scvi_sce.rds") -->

<!-- annotation_colours <- c(annotation_colours, set_names(c("#F4F3EE"), "Unassigned")) -->

<!-- sce_scvi_filt$Line <- str_sub(sce_scvi_filt$sample, end = 6) -->
<!-- sce_scvi_filt$Sample <- str_sub(sce_scvi_filt$sample, start = 8) -->
<!-- sce_scvi_filt$Sample <- case_when( -->
<!--   sce_scvi_filt$Sample == "T" ~ "TIS", -->
<!--   .default = sce_scvi_filt$Sample -->
<!-- ) -->
<!-- sce_scvi_filt$State <- sce_scvi_filt$individual_anno -->

<!-- sce_scvi_filt$State <- case_when( -->
<!--   sce_scvi_filt$State == "Astro" ~ "AC-like", -->
<!--   sce_scvi_filt$State == "Mesenchymal" ~ "MES-like", -->
<!--   sce_scvi_filt$State == "Neuronal" ~ "NPC-like", -->
<!--   sce_scvi_filt$State == "Oligo" ~ "OPC-like", -->
<!--   .default = sce_scvi_filt$State -->
<!-- ) -->


<!-- plots[["bar_anno"]] <- colData(sce_scvi_filt) |> -->
<!--   as_tibble() |> -->
<!--   select(Line, Sample, State) |> -->
<!--   drop_na() |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = Sample, -->
<!--       # y = type, -->
<!--       fill = State -->
<!--     ) -->
<!--   ) + -->
<!--   geom_bar(position = "fill", colour = "black", linewidth = 0.25) + -->
<!--   # geom_bar(stat = "identity") + -->
<!--   facet_grid(~Line) + -->
<!--   scale_fill_manual(values = annotation_colours) + -->
<!--   scale_y_continuous(expand = c(0, 0)) + -->
<!--   theme( -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.major.x = element_blank(), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     strip.background = element_rect(fill = "white", colour = "black"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) + -->
<!--   ylab("Freq") -->

<!-- individual_anno_res <- read_rds( -->
<!--   "/vast/scratch/users/moore.z/pdo_compar/data/intermed/snrnaseq/2_individual_anno_res.rds" -->
<!-- ) -->

<!-- plots[["ridge_anno"]] <- map2( -->
<!--   .x = individual_anno_res$test, -->
<!--   .y = individual_anno_res$res, -->
<!--   .f = function(x, y) { -->
<!--     y |> -->
<!--       mutate(sample = x) |> -->
<!--       select(-c(labels, delta.next, pruned.labels)) |> -->
<!--       pivot_longer(cols = -c(barcode, sample)) -->
<!--   } -->
<!-- ) |> -->
<!--   bind_rows() |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8), -->
<!--     name = case_when( -->
<!--       name == "scores.Astro" ~ "AC-like", -->
<!--       name == "scores.Mesenchymal" ~ "MES-like", -->
<!--       name == "scores.Neuronal" ~ "NPC-like", -->
<!--       name == "scores.Oligo" ~ "OPC-like", -->
<!--       name == "scores.Progenitor" ~ "Progenitor", -->
<!--       name == "scores.Unassigned" ~ "Unassigned" -->
<!--     ) -->
<!--   ) |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   filter(Line != "GL0128") |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = value, -->
<!--       y = name, -->
<!--       fill = Sample -->
<!--     ) -->
<!--   ) + -->
<!--   ggridges::geom_density_ridges2( -->
<!--     panel_scaling = FALSE, -->
<!--     alpha = 0.5, -->
<!--     scale = 0.75, -->
<!--     bandwidth = 0.025, -->
<!--   ) + -->
<!--   facet_wrap(~Line) + -->
<!--   scale_y_discrete(expand = c(0.01, 0.01, 0.125, 0.0), limits = rev) + -->
<!--   scale_fill_manual(values = pie_cols) + -->
<!--   labs(y = "State", x = "Score") + -->
<!--   theme( -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.major.x = element_blank(), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     strip.background = element_rect(fill = "white", colour = "black"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) -->

<!-- patchwork::wrap_plots( -->
<!--   plots$bar_anno, -->
<!--   plots$ridge_anno -->
<!-- ) -->
<!-- # ggsave( -->
<!-- #   filename = "adelaide/anno_plots.pdf", -->
<!-- #   device = cairo_pdf, -->
<!-- #   width = 30, -->
<!-- #   height = 11, -->
<!-- #   units = "cm" -->
<!-- # ) -->
<!-- # EnvironmentModules::module_load("texlive") -->
<!-- # system2( -->
<!-- #   command = "pdfcrop", -->
<!-- #   args = c( -->
<!-- #     "adelaide/anno_plots.pdf", -->
<!-- #     "adelaide/anno_plots.pdf" -->
<!-- #   ), -->
<!-- #   stdout = NULL -->
<!-- # ) -->

<!-- plots[["mde_tumour"]] <- reducedDim(sce_scvi_tumour, "X_scVI_MDE") |> -->
<!--   as_tibble() |> -->
<!--   mutate(bc_well = sce_scvi_tumour$bc_wells) |>  -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!Line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   slice(sample(1:n())) |> -->
<!--   mutate( -->
<!--     State = case_when( -->
<!--       individual_anno == "Astro" ~ "AC-like", -->
<!--       individual_anno == "Mesenchymal" ~ "MES-like", -->
<!--       individual_anno == "Neuronal" ~ "NPC-like", -->
<!--       individual_anno == "Oligo" ~ "OPC-like", -->
<!--       individual_anno == "Unknown" ~ "Unassigned", -->
<!--       .default = individual_anno -->
<!--     ) -->
<!--   ) |> -->
<!--   drop_na() |>  -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = MDE_1, -->
<!--       y = MDE_2, -->
<!--       colour = State, -->
<!--       fill = State, -->
<!--       shape = Sample -->
<!--     ) -->
<!--   ) + -->
<!--   geom_point( -->
<!--     size = 2, -->
<!--     key_glyph = "rect" -->
<!--   ) + -->
<!--   geom_point() +  -->
<!--   scale_colour_manual(values = annotation_colours) + -->
<!--   scale_fill_manual(values = annotation_colours) + -->
<!--   scale_shape_manual(values = type_shapes) + -->
<!--   guides( -->
<!--     # colour = guide_legend(title = "Sample"), -->
<!--     shape = guide_legend(title = "Sample", override.aes = list(size = 3)) -->
<!--   ) + -->
<!--   theme( -->
<!--     # aspect.ratio = 1, -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     axis.text = element_blank(), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) -->

<!-- patchwork::wrap_plots( -->
<!--   plots$bar_anno, -->
<!--   mde_tumour +  -->
<!--       theme( -->
<!--         aspect.ratio = 1,  -->
<!--         legend.position = "none", -->
<!--         plot.title = element_text(vjust=-5), -->
<!--         ) +  -->
<!--       ggtitle("\nTumour Samples - Integrated") -->
<!-- ) -->
<!-- # ggsave( -->
<!-- #   "bar_mde.pdf", -->
<!-- #   device = cairo_pdf, -->
<!-- #   width = 30, -->
<!-- #   height = 11, -->
<!-- #   units = "cm" -->
<!-- # ) -->
<!-- # system2( -->
<!-- #   command = "pdfcrop", -->
<!-- #   args = c( -->
<!-- #     here("bar_mde.pdf"), -->
<!-- #     here("bar_mde.pdf") -->
<!-- #   ), -->
<!-- #   stdout = NULL -->
<!-- # ) -->
<!-- ``` -->



<!-- ```{r norm_plots} -->
<!-- # reducedDim(sce_scvi, "X_scVI_UMAP") <- scater::calculateUMAP( -->
<!-- #   x = sce_scvi, -->
<!-- #   dimred = "X_scVI" -->
<!-- # ) -->
<!-- # patchwork::wrap_plots( -->
<!-- #   scater::plotReducedDim( -->
<!-- #     object = sce_scvi[, !sce_scvi$line == "GL0128"], -->
<!-- #     dimred = "X_scVI_UMAP", -->
<!-- #     colour_by = "normal_res" -->
<!-- #   ), -->
<!-- #   scater::plotReducedDim( -->
<!-- #     object = sce_scvi[, !sce_scvi$line == "GL0128"], -->
<!-- #     dimred = "X_scVI_UMAP", -->
<!-- #     colour_by = "leiden", -->
<!-- #     text_by = "leiden" -->
<!-- #   ) -->
<!-- # ) -->

<!-- plots[["mde_all"]] <- reducedDim(sce_scvi, "X_scVI_MDE") |> -->
<!--   as_tibble(rownames = "bc_well") |> -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate(PTPRC = logcounts(sce_scvi)["ENSG00000081237", ]) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!Line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   slice(sample(1:n())) |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = MDE_1, -->
<!--       y = MDE_2, -->
<!--       colour = Line, -->
<!--       fill = Line, -->
<!--       shape = Sample -->
<!--     ) -->
<!--   ) + -->
<!--   geom_point( -->
<!--     size = 2, -->
<!--     key_glyph = "rect" -->
<!--   ) + -->
<!--   geom_point() + -->
<!--   scale_colour_manual(values = line_cols) + -->
<!--   scale_fill_manual(values = line_cols) + -->
<!--   scale_shape_manual(values = type_shapes) + -->
<!--   guides( -->
<!--     colour = guide_legend(title = "Line"), -->
<!--     shape = guide_legend(title = "Sample", override.aes = list(size = 3)) -->
<!--   ) + -->
<!--   theme( -->
<!--     aspect.ratio = 1, -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     axis.text = element_blank(), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) -->

<!-- plots[["mde_sample"]] <- reducedDim(sce_scvi, "X_scVI_MDE") |> -->
<!--   as_tibble(rownames = "bc_well") |> -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate(PTPRC = logcounts(sce_scvi)["ENSG00000081237", ]) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!Line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   slice(sample(1:n())) |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = MDE_1, -->
<!--       y = MDE_2, -->
<!--       colour = Sample, -->
<!--       # fill = Sample, -->
<!--       shape = Sample -->
<!--     ) -->
<!--   ) + -->
<!--   geom_point( -->
<!--     size = 2, -->
<!--     key_glyph = "rect" -->
<!--   ) + -->
<!--   geom_point() + -->
<!--   # scale_colour_manual(values = line_cols) + -->
<!--   # scale_fill_manual(values = line_cols) + -->
<!--   scale_shape_manual(values = type_shapes) + -->
<!--   guides( -->
<!--     colour = guide_legend(title = "Sample"), -->
<!--     shape = guide_legend(title = "Sample", override.aes = list(size = 3)) -->
<!--   ) + -->
<!--   theme( -->
<!--     aspect.ratio = 1, -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     axis.text = element_blank(), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   )  -->
<!--   # ggh4x::force_panelsizes( -->
<!--   #   rows = unit(10, "cm"), -->
<!--   #   cols = unit(10, "cm") -->
<!--   # ) + -->
<!--   # ggtitle("All Samples Integrated") -->

<!-- plots[["mde_normal"]] <- reducedDim(sce_scvi, "X_scVI_MDE") |> -->
<!--   as_tibble(rownames = "bc_well") |> -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate(PTPRC = logcounts(sce_scvi)["ENSG00000081237", ]) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!Line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   mutate( -->
<!--     Class = case_when( -->
<!--       normal_res == TRUE ~ "Normal", -->
<!--       normal_res == FALSE ~ "Tumour" -->
<!--     ) -->
<!--   ) |> -->
<!--   slice(sample(1:n())) |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = MDE_1, -->
<!--       y = MDE_2, -->
<!--       colour = Class, -->
<!--       fill = Class, -->
<!--       shape = Sample -->
<!--     ) -->
<!--   ) + -->
<!--   geom_point( -->
<!--     size = 2, -->
<!--     key_glyph = "rect" -->
<!--   ) + -->
<!--   geom_point() + -->
<!--   scale_colour_manual(values = c("#77c298", "#e84d60")) + -->
<!--   scale_fill_manual(values = c("#77c298", "#e84d60")) + -->
<!--   scale_shape_manual(values = type_shapes) + -->
<!--   guides( -->
<!--     colour = guide_legend(title = "Class"), -->
<!--     shape = guide_legend(title = "Sample", override.aes = list(size = 3)) -->
<!--   ) + -->
<!--   theme( -->
<!--     # aspect.ratio = 1, -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     axis.text = element_blank(), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) -->

<!-- plots[["mde_ptprc"]] <- reducedDim(sce_scvi, "X_scVI_MDE") |> -->
<!--   as_tibble(rownames = "bc_well") |> -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate(PTPRC = logcounts(sce_scvi)["ENSG00000081237", ]) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!Line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   mutate( -->
<!--     Ellipse = case_when( -->
<!--       leiden == 23 & MDE_1 > 2 ~ "Immune", -->
<!--       # high_oligo == "oligo" ~ "Oligo" -->
<!--     ) -->
<!--   ) |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = MDE_1, -->
<!--       y = MDE_2, -->
<!--       colour = PTPRC, -->
<!--       # fill = Line, -->
<!--       # shape = Sample, -->
<!--       group = Ellipse -->
<!--     ) -->
<!--   ) + -->
<!--   geom_point( -->
<!--     size = 2, -->
<!--     key_glyph = "rect" -->
<!--   ) + -->
<!--   scale_colour_viridis_c() + -->
<!--   ggforce::geom_mark_ellipse( -->
<!--     inherit.aes = FALSE, -->
<!--     mapping = aes( -->
<!--       x = MDE_1, -->
<!--       y = MDE_2, -->
<!--       group = Ellipse, -->
<!--       # colour = Ellipse, -->
<!--       filter = !is.na(Ellipse) -->
<!--     ), -->
<!--     colour = "black", -->
<!--     expand = unit(2, "mm") -->
<!--   ) + -->
<!--   geom_point() + -->
<!--   # scale_colour_manual(values = c("#056517", "#de1a24")) + -->
<!--   # scale_fill_manual(values = c("#056517", "#de1a24")) + -->
<!--   # scale_shape_manual(values = type_shapes) + -->
<!--   guides( -->
<!--     colour = guide_colourbar(title = expression(italic("PTPRC"))), -->
<!--     # shape = guide_legend(title = "Sample", override.aes = list(size = 3)) -->
<!--   ) + -->
<!--   theme( -->
<!--     aspect.ratio = 1, -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     axis.text = element_blank(), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) -->

<!-- plots[["mde_mbp"]] <- reducedDim(sce_scvi, "X_scVI_MDE") |> -->
<!--   as_tibble(rownames = "bc_well") |> -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate(PTPRC = logcounts(sce_scvi)["ENSG00000197971", ]) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   mutate( -->
<!--     Ellipse = case_when( -->
<!--       leiden == 22 & MDE_1 > 2 ~ "Oligo", -->
<!--       # high_oligo == "oligo" ~ "Oligo" -->
<!--     ) -->
<!--   ) |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = MDE_1, -->
<!--       y = MDE_2, -->
<!--       colour = PTPRC, -->
<!--       # fill = Line, -->
<!--       # shape = Sample, -->
<!--       group = Ellipse -->
<!--     ) -->
<!--   ) + -->
<!--   geom_point( -->
<!--     size = 2, -->
<!--     key_glyph = "rect" -->
<!--   ) + -->
<!--   scale_colour_viridis_c() + -->
<!--   ggforce::geom_mark_ellipse( -->
<!--     inherit.aes = FALSE, -->
<!--     mapping = aes( -->
<!--       x = MDE_1, -->
<!--       y = MDE_2, -->
<!--       group = Ellipse, -->
<!--       # colour = Ellipse, -->
<!--       filter = !is.na(Ellipse) -->
<!--     ), -->
<!--     colour = "black", -->
<!--     expand = unit(2, "mm") -->
<!--   ) + -->
<!--   geom_point() + -->
<!--   # scale_colour_manual(values = c("#056517", "#de1a24")) + -->
<!--   # scale_fill_manual(values = c("#056517", "#de1a24")) + -->
<!--   # scale_shape_manual(values = type_shapes) + -->
<!--   guides( -->
<!--     colour = guide_colourbar(title = expression(italic("MBP"))), -->
<!--     # shape = guide_legend(title = "Sample", override.aes = list(size = 3)) -->
<!--   ) + -->
<!--   theme( -->
<!--     aspect.ratio = 1, -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     axis.text = element_blank(), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) -->

<!-- plots[["pie_ptpc"]] <- reducedDim(sce_scvi, "X_scVI_MDE") |> -->
<!--   as_tibble(rownames = "bc_well") |> -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate(PTPRC = logcounts(sce_scvi)["ENSG00000081237", ]) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   mutate( -->
<!--     Ellipse = case_when( -->
<!--       leiden == 23 & MDE_1 > 2 ~ "Immune", -->
<!--       # high_oligo == "oligo" ~ "Oligo" -->
<!--     ) -->
<!--   ) |> -->
<!--   filter(!is.na(Ellipse)) |> -->
<!--   group_by(Sample, Ellipse) |> -->
<!--   count() |> -->
<!--   group_by(Ellipse) |> -->
<!--   mutate(Percent = n / sum(n) * 100) |> -->
<!--   mutate(labels = paste(round(Percent, digits = 2), "%")) |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = "", -->
<!--       y = Percent, -->
<!--       fill = Sample -->
<!--     ) -->
<!--   ) + -->
<!--   scale_fill_manual(values = pie_cols) + -->
<!--   geom_bar(stat = "identity", width = 1) + -->
<!--   coord_polar("y", start = 0) + -->
<!--   geom_text( -->
<!--     aes(label = labels), -->
<!--     position = position_stack(vjust = 0.5), -->
<!--     show.legend = FALSE -->
<!--   ) + -->
<!--   # facet_wrap(~ellipse) + -->
<!--   theme( -->
<!--     aspect.ratio = 1, -->
<!--     axis.title.x = element_blank(), -->
<!--     axis.title.y = element_blank(), -->
<!--     panel.border = element_blank(), -->
<!--     panel.grid = element_blank(), -->
<!--     axis.ticks = element_blank(), -->
<!--     axis.text = element_blank(), -->
<!--     panel.background = element_rect(fill = "white", colour = NA) -->
<!--   ) -->

<!-- plots[["pie_mbp"]] <- reducedDim(sce_scvi, "X_scVI_MDE") |> -->
<!--   as_tibble(rownames = "bc_well") |> -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate(PTPRC = logcounts(sce_scvi)["ENSG00000197971", ]) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   mutate( -->
<!--     Ellipse = case_when( -->
<!--       leiden == 22 & MDE_1 > 2 ~ "Oligo", -->
<!--       # high_oligo == "oligo" ~ "Oligo" -->
<!--     ) -->
<!--   ) |> -->
<!--   filter(!is.na(Ellipse)) |> -->
<!--   group_by(Sample, Ellipse) |> -->
<!--   count() |> -->
<!--   group_by(Ellipse) |> -->
<!--   mutate(Percent = n / sum(n) * 100) |> -->
<!--   mutate(labels = paste(round(Percent, digits = 2), "%")) |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = "", -->
<!--       y = Percent, -->
<!--       fill = Sample -->
<!--     ) -->
<!--   ) + -->
<!--   scale_fill_manual(values = pie_cols) + -->
<!--   geom_bar(stat = "identity", width = 1) + -->
<!--   coord_polar("y", start = 0) + -->
<!--   geom_text( -->
<!--     aes(label = labels), -->
<!--     position = position_stack(vjust = 0.5), -->
<!--     show.legend = FALSE -->
<!--   ) + -->
<!--   # facet_wrap(~ellipse) + -->
<!--   theme( -->
<!--     aspect.ratio = 1, -->
<!--     axis.title.x = element_blank(), -->
<!--     axis.title.y = element_blank(), -->
<!--     panel.border = element_blank(), -->
<!--     panel.grid = element_blank(), -->
<!--     axis.ticks = element_blank(), -->
<!--     axis.text = element_blank(), -->
<!--     panel.background = element_rect(fill = "white", colour = NA) -->
<!--   ) -->

<!-- patchwork::wrap_plots( -->
<!--   patchwork::wrap_plots( -->
<!--     plots$mde_all + ggh4x::force_panelsizes(rows = unit(7.5, "cm"), cols = unit(7.5, "cm")), -->
<!--     plots$mde_normal + theme(aspect.ratio = 1), -->
<!--     patchwork::wrap_plots( -->
<!--       plots$mde_mbp + theme(legend.key.height = unit(0.5, "cm")), -->
<!--       plots$mde_ptprc + theme(legend.key.height = unit(0.5, "cm")), -->
<!--       ncol = 1 -->
<!--     ), -->
<!--     patchwork::wrap_plots( -->
<!--       plots$pie_mbp, -->
<!--       plots$pie_ptpc, -->
<!--       ncol = 1 -->
<!--     ), -->
<!--     ncol = 4 -->
<!--   ), -->
<!--   patchwork::wrap_plots( -->
<!--     plots$mde_tumour + theme(aspect.ratio = 1), -->
<!--     # patchwork::wrap_plots( -->
<!--       plots$bar_anno, -->
<!--       plots$ridge_anno, -->
<!--       # ncol = 2 -->
<!--     # ), -->
<!--     ncol = 3 -->
<!--   ), -->
<!--   patchwork::wrap_plots( -->
<!--   volcano_plots$de_pdn_t, -->
<!--   volcano_plots$de_pdo_t, -->
<!--   rrho_plot, -->
<!--   patchwork::wrap_plots( -->
<!--      plot_enrich$de_pdn_t$up, -->
<!--      plot_enrich$de_pdo_t$up,  -->
<!--      ncol = 1 -->
<!--    ), -->
<!--  nrow = 1 -->
<!-- ), -->
<!--   ncol = 1 -->
<!-- ) + -->
<!--   patchwork::plot_annotation(tag_levels = "A") -->
<!-- ``` -->

<!-- ```{r} -->
<!-- deg_list <- list( -->
<!--   de_pdn_t = read_csv( -->
<!--     file = here("data/processed/broad_pdn_t.csv"), -->
<!--     show_col_types = FALSE -->
<!--   ), -->
<!--   de_pdo_t = read_csv( -->
<!--     file = here("data/processed/broad_pdo_t.csv"), -->
<!--     show_col_types = FALSE -->
<!--   ) -->
<!-- ) -->

<!-- deg_list <- map( -->
<!--   .x = deg_list, -->
<!--   .f = function(x) { -->
<!--     x |> -->
<!--       rename(gene_id = 1) |> -->
<!--       left_join( -->
<!--         y = rowData(sce_scvi_tumour) |> -->
<!--           as_tibble() |> -->
<!--           select(gene_id, gene_name), -->
<!--         by = join_by(gene_id) -->
<!--       ) |> -->
<!--       relocate(gene_name) -->
<!--   } -->
<!-- ) -->

<!-- max_lfc <- map_dbl( -->
<!--   .x = deg_list, -->
<!--   .f = function(x) { -->
<!--     max(x$lfc_mean) -->
<!--   } -->
<!--   ) |> max() -->
<!-- max_bayes <- map_dbl( -->
<!--   .x = deg_list, -->
<!--   .f = function(x) { -->
<!--     max(x$bayes_factor) -->
<!--   } -->
<!--   ) |> max() -->
<!-- min_bayes <- map_dbl( -->
<!--   .x = deg_list, -->
<!--   .f = function(x) { -->
<!--     min(x$bayes_factor) -->
<!--   } -->
<!--   ) |> min() -->

<!-- volcano_plots <- imap( -->
<!--   .x = deg_list, -->
<!--   .f = function(x, i) { -->
<!--     # big boilerplate energy -->
<!--     up_lfc <- x |> -->
<!--       filter(lfc_mean > 2) |> -->
<!--       pull(gene_name) -->
<!--     down_lfc <- x |> -->
<!--       filter(lfc_mean < -2) |> -->
<!--       pull(gene_name) -->
<!--     up_sig_label <- x |> -->
<!--       mutate( -->
<!--         rank = bayes_factor * lfc_mean -->
<!--       ) |> -->
<!--       arrange(-rank) |> -->
<!--       filter(!str_starts(gene_name, "ENSG")) |>  -->
<!--       slice_head(n = 5) |> -->
<!--       pull(gene_name) -->
<!--     down_sig_label <- x |> -->
<!--       mutate( -->
<!--         rank = bayes_factor * lfc_mean -->
<!--       ) |> -->
<!--       arrange(rank) |> -->
<!--       filter(!str_starts(gene_name, "ENSG")) |>  -->
<!--       slice_head(n = 5) |> -->
<!--       pull(gene_name) -->

<!--     x <- x |> -->
<!--       mutate( -->
<!--         sig = case_when( -->
<!--           gene_name %in% up_lfc & is_de_fdr_0.05 == TRUE ~ "up_sig", -->
<!--           gene_name %in% down_lfc & is_de_fdr_0.05 == TRUE ~ "down_sig" -->
<!--         ) -->
<!--       ) |> -->
<!--       mutate( -->
<!--         colour_point = case_when( -->
<!--           sig == "up_sig" ~ "#BC3D41", -->
<!--           sig == "down_sig" ~ "#4F7EBB", -->
<!--           .default = "grey" -->
<!--         ) -->
<!--       ) |> -->
<!--       mutate( -->
<!--         label_point = case_when( -->
<!--           gene_name %in% c(up_sig_label, down_sig_label) ~ gene_name, -->
<!--           .default = NA_character_ -->
<!--         ) -->
<!--       ) -->

<!--     if (i == "de_pdn_t") { -->
<!--       title_up <- "<span style='color:#BC3D41;'>Up in PDN &rarr;</span>" -->
<!--     } else if (i == "de_pdo_t") { -->
<!--       title_up <- "<span style='color:#BC3D41;'>Up in PDO &rarr;</span>" -->
<!--     } -->

<!--     library(ggtext) -->
<!--     # plot -->
<!--     x |> -->
<!--       ggplot( -->
<!--         aes( -->
<!--           x = lfc_mean, -->
<!--           y = bayes_factor, -->
<!--           colour = colour_point, -->
<!--           label = label_point, -->
<!--           shape = is_de_fdr_0.05 -->
<!--         ) -->
<!--       ) + -->
<!--       geom_point(show.legend = FALSE) + -->
<!--       scale_colour_identity() + -->
<!--       labs( -->
<!--         y = "Bayes Factor", -->
<!--         x = expression("log"[2] * "FC") -->
<!--       ) + -->
<!--       theme( -->
<!--         aspect.ratio = 1, -->
<!--         panel.background = element_rect(fill = NA, colour = "black"), -->
<!--         panel.grid = element_line(colour = "grey90"), -->
<!--         # plot.title = element_text(hjust = 0.5) -->
<!--         plot.title = ggtext::element_markdown(hjust = 0.5) -->
<!--       ) + -->
<!--       geom_hline( -->
<!--         yintercept = x |> filter(is_de_fdr_0.05 == FALSE) |> slice(1) |> pull(bayes_factor), -->
<!--         linetype = "dashed" -->
<!--       ) + -->
<!--       geom_vline( -->
<!--         xintercept = c(-2, 2), -->
<!--         linetype = "dashed" -->
<!--       ) + -->
<!--       coord_fixed(clip = "off") + -->
<!--       # ggtitle( -->
<!--       #   "Up in T Up in PDN" -->
<!--       # ) + -->
<!--       annotate( -->
<!--         label = title_up, -->
<!--         geom = "richtext", -->
<!--         colour = "white", -->
<!--         # label.padding = unit(c(0, 0, 0, 0), "lines"), -->
<!--         x = Inf, -->
<!--         y = Inf, -->
<!--         vjust = -0.1, -->
<!--         hjust = 1 -->
<!--       ) + -->
<!--       annotate( -->
<!--         label = "<span style='color:#4F7EBB;;'> &larr; Up in TIS </span>", -->
<!--         geom = "richtext", -->
<!--         colour = "white", -->
<!--         # label.padding = unit(c(0, 0, 0, 0), "lines"), -->
<!--         x = -Inf, -->
<!--         y = Inf, -->
<!--         vjust = -0.1, -->
<!--         hjust = 0 -->
<!--       ) + -->
<!--       labs(title = "", subtitle = "") + -->
<!--       lims( -->
<!--         x = c(-max_lfc, max_lfc), -->
<!--         y = c(min_bayes, max_bayes) -->
<!--       ) +  -->
<!--       ggrepel::geom_label_repel( -->
<!--         max.overlaps = 25, -->
<!--         force = 2, -->
<!--         size = 10 / .pt, -->
<!--         fontface = "italic" -->
<!--       ) -->
<!--     # ggh4x::force_panelsizes( -->
<!--     #   rows = unit(7.5, "cm"), -->
<!--     #   cols = unit(7.5, "cm") -->
<!--     # ) -->
<!--   } -->
<!-- ) -->

<!-- rrho <- RRHO2::RRHO2_initialize( -->
<!--   list1 = deg_list$de_pdn_t |> -->
<!--     # mutate(val = bayes_factor) |> -->
<!--     mutate(val = -log10(bayes_factor) * (lfc_mean)) |> -->
<!--     # filter(group1 == unique(diff$group1)[1]) |> -->
<!--     select(gene_name, val) |> -->
<!--     mutate(val = jitter(val)) |> -->
<!--     distinct(gene_name, .keep_all = TRUE) |> -->
<!--     drop_na() |>  -->
<!--     as.data.frame(), -->
<!--   list2 = deg_list$de_pdo_t |> -->
<!--     # mutate(val = bayes_factor) |> -->
<!--     mutate(val = -log10(bayes_factor) * (lfc_mean)) |> -->
<!--     # filter(group1 == unique(diff$group1)[2]) |> -->
<!--     select(gene_name, val) |> -->
<!--     mutate(val = jitter(val)) |> -->
<!--     distinct(gene_name, .keep_all = TRUE) |> -->
<!--     drop_na() |>  -->
<!--     as.data.frame(), -->
<!--   # BY = TRUE, -->
<!--   # boundary = 0.1 -->
<!--   # log10.ind = TRUE, -->
<!--   # stepsize = 50, -->
<!--   # alternative = "two.sided" -->
<!--   # method = "fisher" -->
<!--   multipleTesting = "none" -->
<!-- ) -->

<!-- RRHO2::RRHO2_heatmap(rrho) -->


<!-- rrho_plot <- lattice::levelplot( -->
<!--   x = rrho$hypermat, -->
<!--   xlab = "pdn vs. tis", -->
<!--   ylab = "pdo vs. tis" -->
<!-- ) -->
<!-- rrho_plot -->

<!-- # names(rrho$hypermat) <- paste("X", 1:ncol(rrho$hypermat)) -->

<!-- hypermat <- rrho$hypermat |> -->
<!--   as_tibble() -->

<!-- rrho_plot <- hypermat |> -->
<!--   rownames_to_column(var = "var1") |> -->
<!--   pivot_longer( -->
<!--     cols = -var1, names_to = "var2", values_to = "value" -->
<!--   ) |> -->
<!--   mutate( -->
<!--     var1 = factor(var1, levels = 1:ncol(rrho$hypermat)), -->
<!--     var2 = factor(gsub("V", "", var2), levels = 1:ncol(rrho$hypermat)) -->
<!--   ) |> -->
<!--   ggplot( -->
<!--     mapping = -->
<!--       aes( -->
<!--         x = var1, -->
<!--         y = var2, -->
<!--         fill = value -->
<!--       ) -->
<!--   ) + -->
<!--   geom_tile() + -->
<!--   scale_fill_viridis_c(na.value = "white") + -->
<!--   labs( -->
<!--     x = "<span style='color:#4F7EBB;'>&larr; Up in TIS</span> -->
<!--         <span style='color:#BC3D41;'>Up in PDN &rarr;</span>", -->
<!--     y = "<span style='color:#4F7EBB;'>&larr; Up in TIS</span> -->
<!--         <span style='color:#BC3D41;'>Up in PDO &rarr;</span>", -->
<!--     fill = expression("-log"[10] * "p") -->
<!--   ) + -->
<!--   theme( -->
<!--     aspect.ratio = 1, -->
<!--     axis.text = element_blank(), -->
<!--     axis.title.y = ggtext::element_markdown(vjust = 0.5), -->
<!--     axis.title.x = ggtext::element_markdown(hjust = 0.5), -->
<!--     axis.ticks = element_blank() -->
<!--   ) -->
<!-- # ggh4x::force_panelsizes( -->
<!-- #   rows = unit(7.5, "cm"), -->
<!-- #   cols = unit(7.5, "cm") -->
<!-- # ) -->

<!-- patchwork::wrap_plots( -->
<!--   volcano_plots$de_pdn_t, -->
<!--   volcano_plots$de_pdo_t, -->
<!--   rrho_plot -->
<!-- ) -->
<!-- # ggsave( -->
<!-- #   filename = here("btrg", "broad_deg.pdf"), -->
<!-- #   device = cairo_pdf, -->
<!-- #   width = 30, -->
<!-- #   height = 50, -->
<!-- #   units = "cm" -->
<!-- # ) -->
<!-- # system2( -->
<!-- #   command = "pdfcrop", -->
<!-- #   args = c( -->
<!-- #     here("btrg", "broad_deg.pdf"), -->
<!-- #     here("btrg", "broad_deg.pdf") -->
<!-- #   ), -->
<!-- #   stdout = NULL -->
<!-- # ) -->

<!-- enrich_broad <- map( -->
<!--   .x = deg_list, -->
<!--   .f = function(x) { -->
<!--     library(enrichR) -->

<!--     up_genes <- x |> -->
<!--       filter(lfc_mean > 0) |> -->
<!--       filter(is_de_fdr_0.05 == TRUE) |> -->
<!--       pull(gene_name) |> -->
<!--       as.character() -->
<!--     down_genes <- x |> -->
<!--       filter(lfc_mean < 0) |> -->
<!--       filter(is_de_fdr_0.05 == TRUE) |> -->
<!--       pull(gene_name) |> -->
<!--       as.character() -->

<!--     up_res <- enrichR::enrichr( -->
<!--       genes = up_genes, -->
<!--       databases = c( -->
<!--         "GO_Molecular_Function_2018", -->
<!--         "KEGG_2013", -->
<!--         "Reactome_2013" -->
<!--       ) -->
<!--     ) -->
<!--     down_res <- enrichR::enrichr( -->
<!--       genes = down_genes, -->
<!--       databases = c( -->
<!--         "GO_Molecular_Function_2018", -->
<!--         "KEGG_2013", -->
<!--         "Reactome_2013" -->
<!--       ) -->
<!--     ) -->
<!--     list( -->
<!--       up = up_res, -->
<!--       down = down_res -->
<!--     ) -->
<!--   } -->
<!-- ) -->

<!-- plot_enrich <- imap( -->
<!--   .x = enrich_broad, -->
<!--   .f = function(x, i) { -->
<!--     if (i == "de_pdn_t") { -->
<!--       title_up <- "<span style='color:#BC3D41;'>Up in PDN</span>" -->
<!--       title_down <- "<span style='color:#4F7EBB;'>Up in TIS</span>" -->
<!--     } else if (i == "de_pdo_t") { -->
<!--       title_up <- "<span style='color:#BC3D41;'>Up in PDO</span>" -->
<!--       title_down <- "<span style='color:#4F7EBB;'>Up in TIS</span>" -->
<!--     } -->

<!--     up_plot <- enrichR::plotEnrich( -->
<!--       df = x$up$GO_Molecular_Function_2018, -->
<!--       orderBy = "Combined.Score", -->
<!--       showTerms = 10 -->
<!--     ) + -->
<!--       scale_fill_viridis_c() + -->
<!--       xlab("Gene Count") +  -->
<!--       guides(fill = guide_colorbar(title = "Score")) + -->
<!--       ggtitle( -->
<!--         title_up -->
<!--       ) + -->
<!--       theme( -->
<!--         # aspect.ratio = 1, -->
<!--         axis.title.y = element_blank(), -->
<!--         plot.title = ggtext::element_markdown() -->
<!--       ) -->
<!--     # ggh4x::force_panelsizes( -->
<!--     #   rows = unit(5, "cm"), -->
<!--     #   cols = unit(10, "cm") -->
<!--     # ) -->
<!--     down_plot <- enrichR::plotEnrich( -->
<!--       df = x$down$GO_Molecular_Function_2018, -->
<!--       orderBy = "Combined.Score", -->
<!--       showTerms = 10 -->
<!--     ) + -->
<!--       scale_fill_viridis_c() + -->
<!--       xlab("Gene Count") +  -->
<!--       guides(fill = guide_colorbar(title = "Score")) + -->
<!--       ggtitle( -->
<!--         title_down -->
<!--       ) + -->
<!--       theme( -->
<!--         # aspect.ratio = 1, -->
<!--         axis.title.y = element_blank(), -->
<!--         plot.title = ggtext::element_markdown() -->
<!--       ) -->
<!--     # ggh4x::force_panelsizes( -->
<!--     #   rows = unit(5, "cm"), -->
<!--     #   cols = unit(10, "cm") -->
<!--     # ) -->
<!--     list(up = up_plot, down = down_plot) -->
<!--   } -->
<!-- ) -->
<!-- patchwork::wrap_plots( -->
<!--   patchwork::wrap_plots( -->
<!--     plot_enrich$de_pdn_t, plot_enrich$de_pdn_t, -->
<!--     ncol = 1 -->
<!--   ), -->
<!--   patchwork::wrap_plots( -->
<!--     plot_enrich$de_pdo_t, plot_enrich$de_pdo_t, -->
<!--     ncol = 1 -->
<!--   ) -->
<!-- ) -->
<!-- # ggsave( -->
<!-- #   filename = here("btrg", "broad_enrich.pdf"), -->
<!-- #   device = cairo_pdf, -->
<!-- #   width = 30, -->
<!-- #   height = 50, -->
<!-- #   units = "cm" -->
<!-- # ) -->
<!-- # system2( -->
<!-- #   command = "pdfcrop", -->
<!-- #   args = c( -->
<!-- #     here("btrg", "broad_enrich.pdf"), -->
<!-- #     here("btrg", "broad_enrich.pdf") -->
<!-- #   ), -->
<!-- #   stdout = NULL -->
<!-- # ) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- cyclone_all <- read_rds( -->
<!--   file = here("data/intermed/snrnaseq/2_cyclone_all.rds") -->
<!-- ) -->
<!-- cyclone_all <- cyclone_all |> -->
<!--   bind_rows() -->

<!-- sce_scvi$phase <- cyclone_all$phases -->

<!-- reducedDim(sce_scvi, "X_scVI_MDE") |> -->
<!--   as_tibble(rownames = "bc_well") |> -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate(PTPRC = logcounts(sce_scvi)["ENSG00000081237", ]) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   mutate( -->
<!--     Ellipse = case_when( -->
<!--       leiden == 23 & MDE_1 > 2 ~ "Immune", -->
<!--       # high_oligo == "oligo" ~ "Oligo" -->
<!--     ) -->
<!--   ) |> -->
<!--   filter(!is.na(Ellipse)) |> -->
<!--   group_by(Sample, phase) |> -->
<!--   count() |> -->
<!--   group_by(Sample) |> -->
<!--   mutate(Percent = n / sum(n) * 100) -->

<!-- reducedDim(sce_scvi, "X_scVI_MDE") |> -->
<!--   as_tibble(rownames = "bc_well") |> -->
<!--   rename(MDE_1 = V1, MDE_2 = V2) |> -->
<!--   left_join(colData(sce_scvi) |> as_tibble(rownames = "bc_well")) |> -->
<!--   mutate(PTPRC = logcounts(sce_scvi)["ENSG00000197971", ]) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   filter(!line == "GL0128") |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   mutate( -->
<!--     Ellipse = case_when( -->
<!--       leiden == 22 & MDE_1 > 2 ~ "Oligo", -->
<!--       # high_oligo == "oligo" ~ "Oligo" -->
<!--     ) -->
<!--   ) |> -->
<!--   filter(!is.na(Ellipse)) |> -->
<!--   group_by(Sample, phase) |> -->
<!--   count() |> -->
<!--   group_by(Sample) |> -->
<!--   mutate(Percent = n / sum(n) * 100) -->

<!-- patchwork::wrap_plots( -->
<!--   mde_all + ggtitle("All Samples Integrated") + -->
<!--     ggh4x::force_panelsizes(rows = unit(10, "cm"), cols = unit(10, "cm")), -->
<!--   mde_sample + ggh4x::force_panelsizes(rows = unit(10, "cm"), cols = unit(10, "cm")) -->
<!-- ) -->


<!-- ggsave( -->
<!--   filename = here("all_compar.pdf"), -->
<!--   device = cairo_pdf, -->
<!--   width = 50, -->
<!--   height = 50, -->
<!--   units = "cm" -->
<!-- ) -->
<!-- system2( -->
<!--   command = "pdfcrop", -->
<!--   args = c( -->
<!--     here("all_compar.pdf"), -->
<!--     here("all_compar.pdf") -->
<!--   ), -->
<!--   stdout = NULL -->
<!-- ) -->

<!-- b <-   patchwork::wrap_plots( -->
<!--     patchwork::wrap_plots( -->
<!--       mde_mbp, -->
<!--       pie_mbp, -->
<!--       mde_ptprc, -->
<!--       pie_ptpc, -->
<!--       nrow = 2 -->
<!--     ) -->
<!--   ) -->

<!-- a <- patchwork::wrap_plots( -->
<!--   mde_all + ggtitle("All Nuclei - Line"), -->
<!--   mde_sample + ggtitle("All Nuclei - Sample"),  -->
<!--   # mde_normal, -->
<!--   # b, -->
<!--   # patchwork::wrap_plots( -->
<!--   #   patchwork::wrap_plots( -->
<!--   #     mde_mbp, -->
<!--   #     pie_mbp, -->
<!--   #     mde_ptprc, -->
<!--   #     pie_ptpc, -->
<!--   #     nrow = 2 -->
<!--   #   ) -->
<!--   # ), -->
<!--   nrow = 1 -->
<!-- ) -->
<!-- ggsave( -->
<!--   plot = a, -->
<!--   filename = here("a.pdf"), -->
<!--   device = cairo_pdf, -->
<!--   width = 25, -->
<!--   height = 10, -->
<!--   units = "cm" -->
<!-- ) -->
<!-- system2( -->
<!--   command = "pdfcrop", -->
<!--   args = c( -->
<!--     here("a.pdf"), -->
<!--     here("a.pdf") -->
<!--   ), -->
<!--   stdout = NULL -->
<!-- ) -->
<!-- patchwork::wrap_plots( -->
<!--   mde_normal + ggtitle("All Nuclei - Class"), -->
<!--   b -->
<!-- ) -->
<!-- ggsave( -->
<!--   filename = here("b.pdf"), -->
<!--   device = cairo_pdf, -->
<!--   width = 26, -->
<!--   height = 10, -->
<!--   units = "cm" -->
<!-- ) -->
<!-- system2( -->
<!--   command = "pdfcrop", -->
<!--   args = c( -->
<!--     here("b.pdf"), -->
<!--     here("b.pdf") -->
<!--   ), -->
<!--   stdout = NULL -->
<!-- ) -->

<!-- ggsave( -->
<!--   mde_normal + ggtitle("All Samples Integrated"), -->
<!--   filename = here("mde_norm.pdf"), -->
<!--   width = 15, -->
<!--   height = 15, -->
<!--   units = "cm" -->
<!-- ) -->
<!-- system2( -->
<!--   command = "pdfcrop", -->
<!--   args = c( -->
<!--     here("mde_norm.pdf"), -->
<!--     here("mde_norm.pdf") -->
<!--   ), -->
<!--   stdout = NULL -->
<!-- ) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- BiocManager::install("zellkonverter", force = TRUE, update = FALSE) -->
<!-- sce_scvi_tumour <- zellkonverter::readH5AD( -->
<!--   file = here("data/intermed/snrnaseq/9_tumour_scvi.h5ad") -->
<!-- ) -->

<!-- assay(sce_scvi_tumour, "norm_counts") <- vroom::vroom( -->
<!--   file = here("data/intermed/snrnaseq/10_tumour_norm_matrix.csv"), -->
<!--   show_col_types = FALSE -->
<!-- ) |> -->
<!--   column_to_rownames(var = 1) |> -->
<!--   as.matrix() -->
<!-- ``` -->

<!-- ```{r individual_anno} -->
<!-- # annotation on integrated norm. sample -->

<!-- # script to load references -->
<!-- source(here("eval/scripts/r/load_ref_list.r")) -->

<!-- # start with refs -->
<!-- # load each reference -->
<!-- ref_list <- load_ref_list() -->

<!-- hvgs <- map( -->
<!--   .progress = TRUE, -->
<!--   .x = ref_list, -->
<!--   .f = function(sce) { -->
<!--     scran::modelGeneVar( -->
<!--       x = sce, -->
<!--       BPPARAM = MulticoreParam( -->
<!--         workers = length(parallelly::availableWorkers()) -->
<!--       ) -->
<!--     ) |> -->
<!--       scran::getTopHVGs(n = 5000) -->
<!--   } -->
<!-- ) -->
<!-- names(hvgs) <- "ref" -->

<!-- # justin -->
<!-- gc() -->

<!-- # could we make this faster? -->
<!-- # embarrassingly parallel -->
<!-- normalised_anno <- map2( -->
<!--   .progress = TRUE, -->
<!--   .x = list(sce_scvi_tumour), -->
<!--   .y = ref_list, -->
<!--   .f = function(test, ref) { -->
<!--     # extract sce -->
<!--     test_sce <- test -->
<!--     ref_sce <- ref -->

<!--     # common genes between -->
<!--     gene_intersect <- intersect( -->
<!--       x = rownames(test_sce), -->
<!--       y = rownames(ref_sce) -->
<!--     ) -->

<!--     # extract hvgs -->
<!--     # hvgs_test <- hvgs$test[[test]] -->
<!--     hvgs_ref <- hvgs$ref -->

<!--     # common hvgs between -->
<!--     hvgs_intersect <- intersect( -->
<!--       x = gene_intersect, -->
<!--       y = hvgs_ref -->
<!--     ) -->

<!--     # column name of labels -->
<!--     # anno <- case_when( -->
<!--     #   ref == "couturier" ~ "cluster", -->
<!--     #   ref == "ruiz_moreno" ~ "annotation_level_3" -->
<!--     # ) -->
<!--     anno <- "cluster" -->

<!--     # run and return -->
<!--     SingleR::SingleR( -->
<!--       test = test_sce[gene_intersect, ], -->
<!--       ref = ref_sce[gene_intersect, ], -->
<!--       restrict = hvgs_intersect, -->
<!--       labels = colData(ref_sce)[, anno], -->
<!--       de.method = "wilcox", -->
<!--       BPPARAM = MulticoreParam( -->
<!--         workers = length(parallelly::availableWorkers()) -->
<!--       ) -->
<!--     ) -->
<!--   } -->
<!-- ) -->

<!-- sce_scvi_tumour$normalised_anno <- normalised_anno[[1]]$pruned.labels -->

<!-- scater::plotReducedDim( -->
<!--   object = sce_scvi_tumour, -->
<!--   dimred = "X_scVI_MDE", -->
<!--   colour_by = "normalised_anno" -->
<!-- ) -->

<!-- sce_scvi_tumour$Line <- str_sub(sce_scvi_tumour$sample, end = 6) -->
<!-- sce_scvi_tumour$Sample <- str_sub(sce_scvi_tumour$sample, start = 8) -->
<!-- sce_scvi_tumour$Sample <- case_when( -->
<!--   sce_scvi_tumour$Sample == "T" ~ "TIS", -->
<!--   .default = sce_scvi_tumour$Sample -->
<!-- ) -->
<!-- sce_scvi_tumour$State <- sce_scvi_tumour$normalised_anno -->

<!-- sce_scvi_tumour$State <- case_when( -->
<!--   sce_scvi_tumour$State == "Astro" ~ "AC-like", -->
<!--   sce_scvi_tumour$State == "Mesenchymal" ~ "MES-like", -->
<!--   sce_scvi_tumour$State == "Neuronal" ~ "NPC-like", -->
<!--   sce_scvi_tumour$State == "Oligo" ~ "OPC-like", -->
<!--   sce_scvi_tumour$State == "Unknown" ~ "Unassigned", -->
<!--   .default = sce_scvi_tumour$State -->
<!-- ) -->

<!-- annotation_colours <- c(annotation_colours, set_names(c("#F4F3EE"), "Unassigned")) -->

<!-- colData(sce_scvi_tumour) |> -->
<!--   as_tibble() |> -->
<!--   select(Line, Sample, State) |> -->
<!--   drop_na() |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = Sample, -->
<!--       # y = type, -->
<!--       fill = State -->
<!--     ) -->
<!--   ) + -->
<!--   geom_bar(position = "fill", colour = "black", linewidth = 0.25) + -->
<!--   # geom_bar(stat = "identity") + -->
<!--   facet_grid(~Line) + -->
<!--   scale_fill_manual(values = annotation_colours) + -->
<!--   scale_y_continuous(expand = c(0, 0)) + -->
<!--   theme( -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.major.x = element_blank(), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     strip.background = element_rect(fill = "white", colour = "black"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) + -->
<!--   ylab("Freq") -->


<!-- plots[["ridge_anno"]] <- map2( -->
<!--   .x = individual_anno_res$test, -->
<!--   .y = individual_anno_res$res, -->
<!--   .f = function(x, y) { -->
<!--     y |> -->
<!--       mutate(sample = x) |> -->
<!--       select(-c(labels, delta.next, pruned.labels)) |> -->
<!--       pivot_longer(cols = -c(barcode, sample)) -->
<!--   } -->
<!-- ) -->


<!-- normalised_anno[[1]] |> -->
<!--   as_tibble() |> -->
<!--   mutate(sample = sce_scvi_tumour$sample) |> -->
<!--   select(-c(labels, delta.next, pruned.labels)) |> -->
<!--   pivot_longer(cols = -c(sample)) |> -->
<!--   mutate( -->
<!--     Line = str_sub(sample, end = 6), -->
<!--     Sample = str_sub(sample, start = 8), -->
<!--     name = case_when( -->
<!--       name == "scores.Astro" ~ "AC-like", -->
<!--       name == "scores.Mesenchymal" ~ "MES-like", -->
<!--       name == "scores.Neuronal" ~ "NPC-like", -->
<!--       name == "scores.Oligo" ~ "OPC-like", -->
<!--       name == "scores.Progenitor" ~ "Progenitor", -->
<!--       name == "scores.Unassigned" ~ "Unassigned" -->
<!--     ) -->
<!--   ) |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   filter(Line != "GL0128") |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = value, -->
<!--       y = name, -->
<!--       fill = Sample -->
<!--     ) -->
<!--   ) + -->
<!--   ggridges::geom_density_ridges2( -->
<!--     panel_scaling = FALSE, -->
<!--     alpha = 0.5, -->
<!--     scale = 0.75, -->
<!--     bandwidth = 0.025, -->
<!--   ) + -->
<!--   facet_wrap(~Line) + -->
<!--   scale_y_discrete(expand = c(0.01, 0.01, 0.125, 0.0), limits = rev) + -->
<!--   scale_fill_manual(values = pie_cols) + -->
<!--   labs(y = "State", x = "Score") + -->
<!--   theme( -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.major.x = element_blank(), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     strip.background = element_rect(fill = "white", colour = "black"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) -->

<!-- bulk_counts <- scuttle::summarizeAssayByGroup( -->
<!--   x = sce_all_labels, -->
<!--   ids = colData(sce_all_labels)$sample, -->
<!--   BPPARAM = MulticoreParam( -->
<!--     workers = length(parallelly::availableWorkers()) -->
<!--   ) -->
<!-- ) -->

<!-- bulk_counts_matrix <- assay(bulk_counts, "sum") |> -->
<!--   as.matrix() -->
<!-- # norm_lib <- edgeR::normLibSizes(bulk_counts_matrix) -->
<!-- # bulk_counts_matrix <- bulk_counts_matrix |> -->
<!-- # apply(MARGIN = 2, FUN = function(x) {x * norm_lib}) -->

<!-- # bulk_counts_matrix <- bulk_counts_matrix * norm_lib -->
<!-- mds <- limma::plotMDS( -->
<!--   x = bulk_counts_matrix[subset_union, ], top = 500, dim.plot = c(1, 2), gene.selection = "common" -->
<!-- ) -->

<!-- dat <- mds$eigen.vectors |> -->
<!--   as_tibble() |> -->
<!--   bind_cols(tibble(sample = colnames(bulk_counts_matrix))) |> -->
<!--   mutate( -->
<!--     line = str_sub(sample, end = 6), -->
<!--     type = str_sub(sample, start = 8) -->
<!--   ) |> -->
<!--   mutate(line = as.factor(line)) -->

<!-- mds_plot <- dat |> -->
<!--   mutate(MDS_1 = V1, MDS_2 = V2, Sample = type, Line = line) |> -->
<!--   mutate(Sample = case_when(Sample == "T" ~ "TIS", .default = Sample)) |> -->
<!--   ggplot( -->
<!--     mapping = aes( -->
<!--       x = MDS_1, -->
<!--       y = MDS_2, -->
<!--       colour = Line, -->
<!--       fill = Line, -->
<!--       shape = Sample -->
<!--     ) -->
<!--   ) + -->
<!--   ggforce::geom_mark_ellipse( -->
<!--     mapping = aes( -->
<!--       group = Line -->
<!--     ), -->
<!--     expand = unit(0, "mm") -->
<!--   ) + -->
<!--   geom_point(size = 5) + -->
<!--   geom_point(key_glyph = draw_key_rect) + -->
<!--   geom_point() + -->
<!--   scale_colour_manual(values = line_cols) + -->
<!--   scale_fill_manual(values = line_cols) + -->
<!--   scale_shape_manual(values = type_shapes) + -->
<!--   xlab(paste0("MDS_1\n", round(mds$var.explained[1] * 100, 2), "% Variance Explained")) + -->
<!--   ylab(paste0("MDS_1\n", round(mds$var.explained[2] * 100, 2), "% Variance Explained")) + -->
<!--   guides( -->
<!--     colour = guide_legend(title = "Line"), -->
<!--     shape = guide_legend(title = "Sample", override.aes = list(size = 3)) -->
<!--   ) + -->
<!--   theme( -->
<!--     aspect.ratio = 1, -->
<!--     axis.ticks = element_line(colour = "black"), -->
<!--     axis.text = element_blank(), -->
<!--     panel.grid.major = element_line(colour = "grey90"), -->
<!--     panel.grid.minor = element_line(colour = "grey90"), -->
<!--     panel.background = element_rect(fill = "white", colour = "black") -->
<!--   ) -->
<!-- ``` -->
